src/main/scala/com/microsoft/hyperspace/actions/RefreshDeleteAction.scala

```diff
@@ -0,0 +1,104 @@
+import com.microsoft.hyperspace.util.ResolverUtils
+
+/**
+ * Refresh index by removing index entries from any deleted source data file.
+ * Note this Refresh Action only fixes an index w.r.t deleted source data files
+ * and does not consider new source data files (if any).
+ * If some original source data file(s) are removed between previous version of index and now,
+ * this Refresh Action updates the index as follows:
+ * 1. Deleted source data files are identified;
+ * 2. Index records' lineage is leveraged to remove any index entry coming
+ *    from those deleted source data files.
+ *
+ * @param spark SparkSession
+ * @param logManager Index LogManager for index being refreshed
+ * @param dataManager Index DataManager for index being refreshed
+ */
+class RefreshDeleteAction(
+    spark: SparkSession,
+    logManager: IndexLogManager,
+    dataManager: IndexDataManager)
+    extends RefreshActionBase(spark, logManager, dataManager) with Logging {
+
+  final override protected def event(appInfo: AppInfo, message: String): HyperspaceEvent = {
+    RefreshDeleteActionEvent(appInfo, logEntry.asInstanceOf[IndexLogEntry], message)
+  }
+
+  /**
+   * For an index with lineage, find all the source data files which have been deleted,
+   * and use index records' lineage to mark and remove index entries which belong to
+   * deleted source data files as those entries are no longer valid.
+   */
+  final override def op(): Unit = {
+    val indexDF = spark.read.parquet(previousIndexLogEntry.content.files.map(_.toString): _*)
+
+    ResolverUtils
+      .resolve(spark, IndexConstants.DATA_FILE_NAME_COLUMN, indexDF.schema.fieldNames) match {
+      case Some(_) =>
+        val deletedFiles = getDeletedFiles
+        logInfo("Refresh index is updating index by removing index entries " +
+          s"from ${deletedFiles.length} deleted source data files.")
+
+        val refreshDF =
+          indexDF.filter(
+            !col(s"${IndexConstants.DATA_FILE_NAME_COLUMN}").isin(deletedFiles: _*))
+
+        refreshDF.write.saveWithBuckets(
+          refreshDF,
+          indexDataPath.toString,
+          previousIndexLogEntry.numBuckets,
+          indexConfig.indexedColumns)
+
+      case None =>
+        throw HyperspaceException(
+          s"Index refresh (to handle deleted source data) is" +
+            " only supported on an index with lineage.")
+    }
+  }
+
+  /**
+   * Compare list of source data files from previous IndexLogEntry to list of
+   * current source data files and identify deleted source data files.
+   *
+   * @return list of full paths of deleted source data files.
+   */
+  private def getDeletedFiles: Seq[String] = {
+    val rels = previousIndexLogEntry.relations
+    val originalFiles = rels.head.data.properties.content.files.map(_.toString)
+    val currentFiles = rels.head.rootPaths.flatMap { p =>
+      Content
+        .fromDirectory(path = new Path(p), throwIfNotExists = true)
```

Comment:
What changed? https://github.com/microsoft/hyperspace/pull/142#discussion_r488333268

Reply:
I tried to explain it in the comment: https://github.com/microsoft/hyperspace/pull/142#discussion_r489103667
I added `throwIfNotExists = true` as current impl does not really let user to delete all source data files and call refresh.
By the time we get to above line (Line 97), some source data file should exist (otherwise it would have failed earlier). That is why I added `throwIfNotExists = true`.
Does this make sense?

Reply:
Yup. I left a comment right before you updated the comment. :)